### ISSUE TITLE
Remove or use normal font for the "Specs" heading

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/ex_doc#602]
Today it is just repetitive and it gives the impression the current section, including the notes about the function definition, are all part of a section named "Specs".

**Comment by josevalim** [comment]
@eksperimental should we remove it or should we use normal font?

**Comment by eksperimental** [comment]
If we remove it we should do something like we do with the functions that has the icon `</>`, and I 'm not sure if it will be clear that those are the specs.

If we leave it, we need to change the font, as that font in normal weight looks bad.

Do we have other examples of documentation where the specs are shown. i would like to see how other people are tackling this issue

**Comment by josevalim** [comment]
> If we leave it, we need to change the font, as that font in normal weight looks bad.

We would use the same found used in the regular text. That's what Erlang docs do too.

**Comment by josevalim** [comment]
> Do we have other examples of documentation where the specs are shown. i would like to see how other people are tackling this issue

http://erlang.org/doc/man/lists.html#all-2

**Comment by eksperimental** [comment]
Playing with it, i think it's better to get rid of the colored background, 
![ex_doc_specs](https://cloud.githubusercontent.com/assets/9133420/18755224/86d0b4a2-8115-11e6-844f-498260207dd2.png)

"Specs" uses same font like the spec code.

(it's just a sketch, i'm horrible with design) :-p

**Comment by josevalim** [comment]
@eksperimental I think Specs should either use the same font as all other headings (how it works today) OR it should be the regular text font. I don't think we should introduce yet another font/demarcation/heading. Removing the background also looks neat to me but I will delegate the decision to @dignifiedquire :)

**Comment by eksperimental** [comment]
Regular font, looks wrong (size too big)

this is regular font, size: 0.90em
![image](https://cloud.githubusercontent.com/assets/9133420/18755803/bb614a5e-8117-11e6-9e9f-295900555736.png)

HEADING FONT (bold)
![image](https://cloud.githubusercontent.com/assets/9133420/18755993/63c3480a-8118-11e6-8e6d-f0919293cea5.png)

HEADING FONT (normal weight)
![image](https://cloud.githubusercontent.com/assets/9133420/18756030/8b1a5e66-8118-11e6-94f1-61e4bf0caace.png)

I would go with the last one.

**Comment by bbugh** [comment]
The Elixir docs are *wonderful*, especially in comparison to most languages, and I've noticed myself regularly struggling to remember function signatures. Having the specs there is helpful, but I think it is harder to remember because the type info is out of context of the signature.

Many languages use the types as parameter names (`concat` example: [Ruby](http://ruby-doc.org/core-1.9.3/String.html#method-i-concat), [Erlang](http://erlang.org/doc/man/string.html#concat-2), [Elm](http://package.elm-lang.org/packages/elm-lang/core/4.0.5/String#concat)), but ex_doc uses the friendly name, which is nice, but I think some variation where they're either in context above the signature or integrated into signature would make it more memorable and clear. 

I like the idea of having them integrated into the function signature like other languages do. Dialyzer seems to get pretty verbose sometimes though, so I'm not sure how it would work in practice. Here's a few examples (accidentally using `Enumerable.t` instead of `Enum.t`, sorry!) 

*#*1 Without variable names (my personal favorite):
![image](https://cloud.githubusercontent.com/assets/438465/18834453/917e9920-83bb-11e6-9953-d8a157b74f24.png) 

*#*2 With variable names on a short method:
![image](https://cloud.githubusercontent.com/assets/438465/18834550/33f5b88c-83bc-11e6-9b1b-366a8607d451.png)

*#*3 With variables names, with types less bold:
![image](https://cloud.githubusercontent.com/assets/438465/18834627/9e473954-83bc-11e6-949d-b213308ac6ae.png)

*#*4 Longer functions, only types
![image](https://cloud.githubusercontent.com/assets/438465/18835576/4ecc0eae-83c1-11e6-8acc-be5d47299010.png)

*#*5 Longer function with both names and types:
![image](https://cloud.githubusercontent.com/assets/438465/18835659/d4befcce-83c1-11e6-89ac-e69a8be971cc.png)

*#*6 Here's a (not beautiful) example of the spec being above the signature like it is in the code. 
![image](https://cloud.githubusercontent.com/assets/438465/18833397/ff8db7b2-83b5-11e6-8a63-785e69553109.png)

Finally, the future `concat` example for Elixir? 😸 

![image](https://cloud.githubusercontent.com/assets/438465/18835321/26aabaf2-83c0-11e6-8dfb-c2d49c7afe33.png)

**Comment by josevalim** [comment]
I actually like 6. It also brings more attention to the function header altogether, which is a common complaint. I would put it *under* the signature.

I don't fix we can mix function definition with the specs though. Sometimes specs have multiple clauses, other times they have their own variables too.

**Comment by josevalim** [comment]
And thank you for the feedback @bbugh!

**Comment by dignifiedquire** [comment]
It is interesting there is such a discussion around this, as this was one of the sections I had the hardest time to design. While I generally like the ideas from @bbugh and agree with the sentiment, I don't think it would be good to have the function headers vary in their height. 

The hard part about this section is that it can regularly be just a single spec, in which case it would make so much sense to merge it as @bbugh suggested, but it also can a list of many as @josevalim pointed out. 

Generally I would be in favour of dropping the title `Specs`, as it doesn't help much in the current state. In addition, without the title there might be some more room to improve on the background/highlighting to make clearer that this describes the above defined function. I will think on that a bit more.

**Comment by eksperimental** [comment]
I was going to propose something like number 6, but couldn't design it.
it should be the specs before the signature, and the bar in the left with a lighter colour, and maybe lighter background color aswell.

**Comment by josevalim** [comment]
To continue moving the discussion here, here is a proposal for 6 where:

  1. The specs comes after the signature (but with opaque 0.6)
  2. The inside of the function is indented (1.2em)
  3. The space between the function contents and headers are increased (from 0 to 1.2em)

<img width="1269" alt="screen shot 2016-09-28 at 13 04 31" src="https://cloud.githubusercontent.com/assets/9582/18911263/5c5c0a94-857c-11e6-8ab6-80e2bb3dfd00.png">

**Comment by josevalim** [comment]
I have also added a title to the typespecs section so a short explanation appears on mouse hover but we cannot see it on the screenshot.

**Comment by eksperimental** [comment]
@josevalim that's beautiful, and that's what I meant in my last message (I expressed it backwards specs before signature)... but that's exactly what I meant!
Could you please send a screenshot of how it looks with a multi-line spec?

If it looks ok with multi-line spec i will just go with this one, and it conveys the message and it has brings zero distraction to the table.

thank you!

**Comment by josevalim** [comment]
@eksperimental can you find a multi-line spec? :D

Also, note to @josevalim: also tidy up the types section once this is approved.

**Comment by josevalim** [comment]
Ok, I found a multi-line one:

<img width="1266" alt="screen shot 2016-09-28 at 15 56 47" src="https://cloud.githubusercontent.com/assets/9582/18916288/3e2e91be-8594-11e6-8685-d534e631698b.png">

**Comment by josevalim** [comment]
Here are two other approaches. One makes the border and the whole bottom of the heading more opaque:

<img width="1251" alt="screen shot 2016-09-28 at 16 03 28" src="https://cloud.githubusercontent.com/assets/9582/18916642/5cf3ea9e-8595-11e6-9459-01cc79256ea1.png">

This keeps specs in its own section without background:

<img width="1257" alt="screen shot 2016-09-28 at 16 04 45" src="https://cloud.githubusercontent.com/assets/9582/18916651/654bad76-8595-11e6-9f5a-03746c2d5755.png">

Thoughts?

**Comment by dignifiedquire** [comment]
I really like the last one, but fear that with the single line ones (which as far as I can tell are much more common) it gets lost easily.

**Comment by josevalim** [comment]
@dignifiedquire yup, I am afraid the last one puts us back where we started.

**Comment by bbugh** [comment]
For what it's worth, I like either the first or the second one. 

For what it's worth, I'm strongly in favor of the spec being in the same context as the function. I think your first example @josevalim is what I would pick personally. 😄 Here it is in dark mode:

![image](https://cloud.githubusercontent.com/assets/438465/18916974/fcb1bbb8-855b-11e6-8ea9-cabba04ca225.png)

Another rendition with the primary function signature in slightly bigger text and a different background color for differentiation

![image](https://cloud.githubusercontent.com/assets/438465/18917388/633de5e0-855d-11e6-844d-ac39cbb6bf68.png)

**Comment by eksperimental** [comment]
I like Jose's second example.
playing with gimp I have added a bottom border, and reduced the top padding of the spec (only on B and C)

I think the signature's bottom border draws a line (literally) between the signature and the spec.
So far I like B the best

A: Bars of same color, opaque background for spec
B: Different colors in bars, + bottom border for signature. Top padding in spec reduced. 
<img width="1251" alt="exdoc_a_b" src="https://cloud.githubusercontent.com/assets/9133420/18919544/3703a5f8-8564-11e6-8738-ffc2fcdd1a29.png">

C: Bars of same color, + bottom border for signature. Top padding in spec reduced. 
<img width="1251" alt="exdoc_c" src="https://cloud.githubusercontent.com/assets/9133420/18919564/4465ee54-8564-11e6-96f7-a12ec7e4c1ba.png">

**Comment by josevalim** [comment]
I have pushed the first option to master so people have something more accessible to play with: http://elixir-lang.org/docs/master/logger/Logger.html

I will continue to take feedback and improving it. I have also unified how we handle types with the remaining of the functions/callbacks. Compare how types work now on [master](http://elixir-lang.org/docs/master/logger/Logger.html#types) with [stable](http://elixir-lang.org/docs/stable/logger/Logger.html#types).

**Comment by michalmuskala** [comment]
It seems there's a bug where if the function name is the same as a type name it will be highlighted and linked to the type http://elixir-lang.org/docs/master/logger/Logger.html#level/0

<img width="229" alt="screenshot 2016-09-28 19 18 57" src="https://cloud.githubusercontent.com/assets/477062/18924432/759b47c0-85b0-11e6-8588-0f1ccc70f3ad.png">

**Comment by josevalim** [comment]
@michalmuskala that bug exists before this patch it seems. Please open up a new issue?

**Comment by josevalim** [comment]
This was approved by @dignifiedquire, so I am closing it.

**Comment by Odaeus** [comment]
May I ask what happened to this proposal? The last comment doesn't quite make sense to me, it was approved but nothing changed? I was about to request feedback on whether we could improve the "Specs" heading and found this old issue which matches my thoughts exactly!

**Comment by eksperimental** [comment]
> May I ask what happened to this proposal? The last comment doesn't quite make sense to me, it was approved but nothing changed? I was about to request feedback on whether we could improve the "Specs" heading and found this old issue which matches my thoughts exactly!

Hi. The issue is 5 years old. A lot has changed, and without a copy of the html at the time that this proposal was done it is really hard to understand what is being talk about in this thread.

In which ways you think the specs could be improved.

**Comment by Odaeus** [comment]
Hi @eksperimental, what I mean is that the "Specs" header and definition appears unchanged since this issue was raised. Perhaps the HTML is different but the visual display looks identical to me. Admittedly, I don't think there is a screenshot of the style at the time, and I am inferring this from the comments and mock-ups.

I agree with the problem statement in the very first comment that started this issue. Plus I'd add that I find myself often reading the type spec and momentarily being confused due to the visual hierarchy, because I actually want the names of the function arguments. Before I found this issue I created a suggested mock-up that matches some of the suggestions above:

Current:
![CleanShot 2022-01-02 at 18 12 06](https://user-images.githubusercontent.com/2150/147884771-dfe6adef-ebc0-49e8-b4d3-2a42046dd5a9.png)

Proposal:
![CleanShot 2022-01-02 at 18 11 46](https://user-images.githubusercontent.com/2150/147884777-de65eab7-eb36-494d-a2b0-b4cbf113079e.png)

I was going to post this on the Elixir forum first to get feedback on what people find easier to read.

I just wanted to know if anyone can remember what happened here 😄 , because I couldn't determine why this issue was closed and didn't want to waste my time proposing something that had been rejected years ago.

**Comment by eksperimental** [comment]
I agree. It is confusing. It makes sense for documentation of types, but not for functions.
Maybe something like this?
![image](https://user-images.githubusercontent.com/9133420/147885359-3976d93a-335f-4163-b82a-3a33cd3e50b0.png)

**Comment by Odaeus** [comment]
@eksperimental it's definitely an improvement as it conveys that the heading and the content are grouped together now. So I'd be really happy to see that change as it solves my main complaint!

Optionally, I'd argue that the word "Specs" itself is superfluous as I think the convention of monospaced typeface, no background (unlike code), and placed underneath the function signature is strong enough to distinguish typespecs without confusion. I'm not sure if this would negatively affect accessibility.

**Comment by eksperimental** [comment]
We could maybe remove the "Spec" header, and use `@spec map_join(...) :: ...` the same way IEx shows the specs

**Comment by Odaeus** [comment]
That has a lot going for it! It's low-key, descriptive, and consistent with defining a spec and IEx output. I think it's great.

I did some quick mock-ups as I noticed the IEx output has the "@spec" keyword coloured, which I think makes sense because this will be a repeating element. It makes it quicker to identify and visually skip it if there's a slight distinction. I know syntax highlighting doesn't apply to the specs line, so I guess it would be set in the stylesheet.

Plain, I don't think this is terrible:
![CleanShot 2022-01-03 at 10 15 00](https://user-images.githubusercontent.com/2150/147915670-7e603452-2ef4-4344-b1f6-382597e62dc2.png)

Same as syntax highlighter, perhaps too strong:
![CleanShot 2022-01-03 at 10 16 40](https://user-images.githubusercontent.com/2150/147915698-efe23f37-4a92-4ac7-8622-06b5858a8ee2.png)

Grey, I think this looks good:
![CleanShot 2022-01-03 at 10 31 40](https://user-images.githubusercontent.com/2150/147916087-4bd06bdc-e277-4e6d-9efe-eb15087229c3.png)

Let me know what you think!

**Comment by eksperimental** [comment]
> That has a lot going for it! It's low-key, descriptive, and consistent with defining a spec and IEx output. I think it's great.
> 
> I did some quick mock-ups as I noticed the IEx output has the "@SPEC" keyword coloured, which I think makes sense because this will be a repeating element. It makes it quicker to identify and visually skip it if there's a slight distinction. I know syntax highlighting doesn't apply to the specs line, so I guess it would be set in the stylesheet.
> 
> Plain, I don't think this is terrible: ![CleanShot 2022-01-03 at 10 15 00](https://user-images.githubusercontent.com/2150/147915670-7e603452-2ef4-4344-b1f6-382597e62dc2.png)
> 
> Same as syntax highlighter, perhaps too strong: ![CleanShot 2022-01-03 at 10 16 40](https://user-images.githubusercontent.com/2150/147915698-efe23f37-4a92-4ac7-8622-06b5858a8ee2.png)
> 
> Grey, I think this looks good: ![CleanShot 2022-01-03 at 10 31 40](https://user-images.githubusercontent.com/2150/147916087-4bd06bdc-e277-4e6d-9efe-eb15087229c3.png)
> 
> Let me know what you think!

Grey looks beautiful. I really like this one addition.
It removes the ambiguity and brings in consistency.

**Comment by eksperimental** [comment]
@Odaeus feel free to send a PR if you can.
